fem/tmop.cpp

```diff
@@ -957,42 +957,181 @@ void AnalyticAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
 }
 
 #ifdef MFEM_USE_MPI
-void DiscreteAdaptTC::SetParDiscreteTargetSpec(ParGridFunction &tspec_)
+void DiscreteAdaptTC::FinalizeParDiscreteTargetSpec()
 {
-   tspec.SetSize(tspec_.Size());
-   tspec = tspec_;
-   tspec_fes   = tspec_.FESpace();
+   MFEM_VERIFY(adapt_eval, "SetAdaptivityEvaluator() has not been called!")
+   MFEM_VERIFY(ncomp > 0, "No target specifications have been set!");
 
-   if (!adapt_eval) { MFEM_ABORT("Set adaptivity evaluator\n"); }
+   adapt_eval->SetParMetaInfo(*ptspec_fes->GetParMesh(),
+                              *ptspec_fes->FEColl(),
+                              ncomp);
+   adapt_eval->SetInitialField(*tspec_fes->GetMesh()->GetNodes(), tspec);
 
-   adapt_eval->SetParMetaInfo(*tspec_.ParFESpace()->GetParMesh(),
-                              *tspec_.FESpace()->FEColl(),
-                              tspec_.FESpace()->GetVDim());
+   tspec_sav = tspec;
+   tspec_fesv = new FiniteElementSpace(tspec_fes->GetMesh(),
+                                       tspec_fes->FEColl(),
+                                       ncomp);
+}
 
-   adapt_eval->SetInitialField
-   (*tspec_.FESpace()->GetMesh()->GetNodes(), tspec);
+void DiscreteAdaptTC::SetParDiscreteTargetBase(ParGridFunction &tspec_)
+{
+   const int vdim = tspec_.FESpace()->GetVDim(),
+             cnt = tspec_.Size()/vdim;
 
-   tspec_sav = tspec;
+   ncomp += vdim;
+
+   if (tspec_fes == NULL)
+   {
+      tspec_fes = new FiniteElementSpace(tspec_.FESpace()->GetMesh(),
+                                         tspec_.FESpace()->FEColl(),
+                                         1);
+      ptspec_fes = tspec_.ParFESpace();
+      tspec = tspec_;
+
+      return;
+   }
+
+   // need to append data to tspec
+   // make a copy of tspec->tspec_temp, increase its size, and
+   // copy data from tspec_temp -> tspec, then add new entries
+   Vector tspec_temp = tspec;
+   tspec.SetSize(ncomp*cnt);
+
+   for (int i = 0; i < tspec_temp.Size(); i++)
+   {
+      tspec(i) = tspec_temp(i);
+   }
+
+   for (int i = 0; i < cnt*vdim; i++)
+   {
+      tspec(i+(ncomp-vdim)*cnt) = tspec_(i);
+   }
+}
+
+void DiscreteAdaptTC::SetParDiscreteTargetSize(ParGridFunction &tspec_)
+{
+   MFEM_VERIFY(sizeidx == -1, " Size discrete function already specified")
```

Comment:
```suggestion
   if (sizeidx > -1) { SetTspecAtIndex(sizeidx, stpec_); }
```
This would allow to call these functions after Finalize() to update the targets.